File: lib/detect.js
/**
 * Detect existing AI instruction files in the project root.
 * Returns an array of { tool, file, fullPath, isDir } objects.
 */

const fs = require("fs");
const path = require("path");

const INSTRUCTION_FILES = [
  { tool: "Claude Code", file: "CLAUDE.md" },
  { tool: "Cursor", file: ".cursorrules" },
  { tool: "Cursor", file: ".cursor/rules" },
  { tool: "GitHub Copilot", file: ".github/copilot-instructions.md" },
  { tool: "Windsurf", file: ".windsurfrules" },
  { tool: "Windsurf", file: ".windsurf/rules" },
  { tool: "Cline", file: ".clinerules" },
  { tool: "Codex / Gemini", file: "AGENTS.md" },
  { tool: "Antigravity", file: ".antigravity/rules.md" },
  { tool: "Gemini CLI", file: ".gemini/GEMINI.md" },
  { tool: "Aider", file: "CONVENTIONS.md" },
];

/**
 * Find AGENTS.md case-insensitively in projectRoot.
 * Returns the actual filename (e.g. "agents.md", "Agents.md") or null.
 */
function findAgentsMdCaseInsensitive(projectRoot) {
  try {
    const entries = fs.readdirSync(projectRoot);
    const match = entries.find(
      (e) => e.toLowerCase() === "agents.md" && fs.statSync(path.join(projectRoot, e)).isFile()
    );
    return match || null;
  } catch {
    return null;
  }
}

function detectInstructionFiles(projectRoot) {
  const found = [];
  const seenLower = new Set();

  for (const entry of INSTRUCTION_FILES) {
    // Skip AGENTS.md in the static list — handled by case-insensitive scan
    if (entry.file.toLowerCase() === "agents.md") continue;

    const fullPath = path.join(projectRoot, entry.file);
    if (fs.existsSync(fullPath)) {
      const stat = fs.statSync(fullPath);
      found.push({
        tool: entry.tool,
        file: entry.file,
        fullPath,
        isDir: stat.isDirectory(),
      });
      seenLower.add(entry.file.toLowerCase());
    }
  }

  // Case-insensitive AGENTS.md detection
  const agentsFile = findAgentsMdCaseInsensitive(projectRoot);
  if (agentsFile && !seenLower.has(agentsFile.toLowerCase())) {
    found.push({
      tool: "Codex / Gemini",
      file: agentsFile,
      fullPath: path.join(projectRoot, agentsFile),
      isDir: false,
    });
  }

  return found;
}

module.exports = { detectInstructionFiles, findAgentsMdCaseInsensitive, INSTRUCTION_FILES };

File: rust/src/detect.rs
use std::fs;
use std::path::{Path, PathBuf};

pub struct InstructionFile {
    pub tool: &'static str,
    pub file: &'static str,
    pub full_path: PathBuf,
    pub is_dir: bool,
}

const INSTRUCTION_FILES: &[(&str, &str)] = &[
    ("Claude Code", "CLAUDE.md"),
    ("Cursor", ".cursorrules"),
    ("Cursor", ".cursor/rules"),
    ("GitHub Copilot", ".github/copilot-instructions.md"),
    ("Windsurf", ".windsurfrules"),
    ("Windsurf", ".windsurf/rules"),
    ("Cline", ".clinerules"),
    ("Antigravity", ".antigravity/rules.md"),
    ("Gemini CLI", ".gemini/GEMINI.md"),
    ("Aider", "CONVENTIONS.md"),
];

/// Case-insensitive search for AGENTS.md
pub fn find_agents_md(project_root: &Path) -> Option<String> {
    let entries = fs::read_dir(project_root).ok()?;
    for entry in entries.flatten() {
        let name = entry.file_name();
        let name_str = name.to_string_lossy();
        if name_str.to_lowercase() == "agents.md" {
            return Some(name_str.into_owned());
        }
    }
    None
}

/// Detect all existing instruction files in project root
pub fn detect_instruction_files(project_root: &Path) -> Vec<InstructionFile> {
    let mut found = Vec::new();

    // Check AGENTS.md first (case-insensitive)
    if let Some(agents_name) = find_agents_md(project_root) {
        let full_path = project_root.join(&agents_name);
        found.push(InstructionFile {
            tool: "Codex / Gemini",
            file: Box::leak(agents_name.into_boxed_str()),
            full_path,
            is_dir: false,
        });
    }

    // Check static list
    for &(tool, file) in INSTRUCTION_FILES {
        let full_path = project_root.join(file);
        if full_path.exists() {
            // Skip if it's AGENTS.md (already found above)
            if file.to_lowercase() == "agents.md" {
                continue;
            }
            let is_dir = full_path.is_dir();
            found.push(InstructionFile {
                tool,
                file,
                full_path,
                is_dir,
            });
        }
    }

    found
}
